# Applied Data Science Capstone Assignment 2 
## Segmenting and Clustering Neighborhoods in Toronto
### Battle of the Neighbourhoods - Toronto 
Author: Alex Pr <br>
This notebook contains Questions 1, 2 , 3 of the Assignment.

# Question 1
### Importing Data

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

In [32]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


In [33]:
# Import necessary libraries
import pandas as pd
import requests
import lxml.html as lh

In [34]:
# given link
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [35]:
#Create page to handle the contents of the website
page = requests.get(url)
#Store the contents of the website
wiki = pd.read_html(page.text)
print('length wiki:', len(wiki))
print(wiki) # we see the requred data only in first element

length wiki: 3
[    Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                       

In [36]:
print(wiki[0])

    Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       

In [37]:
#Creating the pandas data frame
df0=pd.DataFrame(wiki[0])
df0

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [38]:
# rename columns
df0.columns = ['Postcode','Borough', 'Neighbourhood']
# remove "Not assigned"
df = df0[df0["Borough"] != "Not assigned"]
df = df.reset_index(drop=True)
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A  Downtown Toronto   
..       ...               ...   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [39]:
# Chek if we have cleaned all "Not assigned"
print('"Not assigned" values =', df[df['Neighbourhood'] == 'Not assigned'].count().sum())

"Not assigned" values = 0


In [40]:
#save our df to file
df.to_csv(r'df_canada.csv')
#check the answer
df.shape

(103, 3)

### Answer to Question 1: DataFrame has 103 rows and 3 columns

# Question 2
Installing Geocoder and use the Geocoder package to create the dataframe with coordinates.

In [41]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [42]:
import geocoder

The code below was tried to run. Now changed this field to text. <br>


lat_lng_coords = None
postal_code = 'M3A'

loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

It seems to be too slow. Let's get data from a link.

In [43]:
# Get geographical coordinates of each postal code
link = "http://cocl.us/Geospatial_data"
df2 = pd.read_csv(link)
print('df Neighborhoods shape is', df.shape)
print('df Coordinates shape is', df2.shape)
df2.head()

df Neighborhoods shape is (103, 3)
df Coordinates shape is (103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We see that the shape of df3 equals the shape of our dataframe for Neighborhoods. But we need to change column name Postal code to Postcode to merge two data frames together.

In [44]:
df2.columns = ['Postcode','Latitude','Longitude']
df_new = pd.merge(df, df2, on='Postcode')
df_new.sort_values(by=['Postcode'], inplace=True)
df_new = df_new.reset_index(drop=True)
df_new.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Malvern, Rouge  43.806686   
1      M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [45]:
df_Toronto = df_new
df_Toronto.shape

(103, 5)

### Answer to Question 2: DataFrame with Coordinates has 103 rows and 5 columns

# Question 3
Explore and cluster the neighborhoods in Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [46]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [47]:
import numpy as np
import json
from pandas.io.json import json_normalize
from geopy.geocoders import Nominatim
#from geopy.geocoders import Nominatim
import folium
from sklearn.cluster import KMeans

import matplotlib.cm as cm
import matplotlib.colors as colors
print('Libraries imported.')

Libraries imported.


## Map of Toronto

In [48]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [49]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_Toronto['Latitude'], df_Toronto['Longitude'], df_Toronto['Borough'], df_Toronto['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Make this Notebook Trusted to load map

Using the Foursquare API to explore the neighborhoods and segment them.

In [50]:
CLIENT_ID = 'UGHOCJQLX4ZG2WEWLDFNRHU3W5XXHLBQALKX1BZBUIZCFTWZ' # Foursquare ID
CLIENT_SECRET = 'HQFIPE1IU3SYVK0YRNT1KW0D3UOLST3YCKSOS1TUWRG3FKXD' # Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

## Explore Neighborhoods in Toronto

 Function to get all the venue categories in Toronto

In [51]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [52]:
# run the above function on each neighborhood and create a new dataframe called toronto_venues.
toronto_venues = getNearbyVenues(names=df_Toronto['Neighbourhood'],
                                   latitudes=df_Toronto['Latitude'],
                                   longitudes=df_Toronto['Longitude'],
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

In [53]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  \
0                          Malvern, Rouge               43.806686   
1  Rouge Hill, Port Union, Highland Creek               43.784535   
2  Rouge Hill, Port Union, Highland Creek               43.784535   
3       Guildwood, Morningside, West Hill               43.763573   
4       Guildwood, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                                              Venue  \
0               -79.194353                                            Wendy’s   
1               -79.160497                              Royal Canadian Legion   
2               -79.160497  SEBS Engineering Inc. (Sustainable Energy and ...   
3               -79.188711                                     RBC Royal Bank   
4               -79.188711                                  G & G Electronics   

               Venue Category  
0        Fast Food Restaurant  
1                         Bar  
2  Construction & Landscaping  
3                        Bank  
4           Electronics Store

In [54]:
toronto_venues.shape

(1338, 5)

Let's check how many venues were returned for each neighbourhood

In [55]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                4   
Woodbine Heights                                                      9   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              5   
Alderwood, Long Branch                                                 7   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     22   
...                                                                  ...   
Willowdale, Willowdale West                                            5   
Woburn                                                                 4   
Woodbine Heights                                                       9   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Category  
Neighbourhood                                                           
Agincourt                                            5               5  
Alderwood, Long Branch                               7               7  
Bathurst Manor, Wilson Heights, Downsview North     21              21  
Bayview Village                                      4               4  
Bedford Park, Lawrence Manor East                   22              22  
...                                                ...             ...  
Willowdale, Willowdale West                          5               5  
Woburn                                               4               4  
Woodbine Heights                                     9               9  
York Mills West                                      2               2  
York Mills, Silver Hills                             1               1  

[96 rows x 4 columns]

In [56]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 240 uniques categories.


### Analyze Each Neighborhood

In [57]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  \
0                          Malvern, Rouge                  0        0   
1  Rouge Hill, Port Union, Highland Creek                  0        0   
2  Rouge Hill, Port Union, Highland Creek                  0        0   
3       Guildwood, Morningside, West Hill                  0        0   
4       Guildwood, Morningside, West Hill                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0                 0                    0             0  ...              0   
1                 0                    0             0  ...              0   
2                 0                    0             0  ...              0   
3                 0                    0             0  ...              0   
4                 0                    0             0  ...              0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 241 columns]

In [58]:
# Examine the new data frame
toronto_onehot.shape

(1338, 241)

Group rows by neighbourhood and by taking the mean of the occurrence frequency of each category

In [59]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.shape

(96, 241)

Look at each neighbourhood along with the top 5 most common venues

In [60]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Breakfast Spot   0.2
1                     Lounge   0.2
2  Latin American Restaurant   0.2
3             Clothing Store   0.2
4               Skating Rink   0.2


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.29
1  Sandwich Place  0.14
2             Gym  0.14
3        Pharmacy  0.14
4             Pub  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
         venue  freq
0         Bank  0.10
1  Coffee Shop  0.10
2     Pharmacy  0.05
3   Restaurant  0.05
4  Pizza Place  0.05


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1                 Café  0.25
2   Chinese Restaurant  0.25
3                 Bank  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.09
1         Coffee Shop  0.09
2  Italian Restaurant  0.09
3         Pizza Place  0.05
4   Indian Restaurant  0.05

Sorting the venues in descending order.

In [61]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighbourhood.

In [62]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head(10)

Neighbourhood 1st Most Common Venue  \
0                                          Agincourt                Lounge   
1                             Alderwood, Long Branch           Pizza Place   
2    Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                    Bayview Village                  Café   
4                  Bedford Park, Lawrence Manor East           Coffee Shop   
5                                        Berczy Park           Coffee Shop   
6                        Birch Cliff, Cliffside West       College Stadium   
7       Brockton, Parkdale Village, Exhibition Place                  Café   
8  Business reply mail Processing Centre, South C...      Recording Studio   
9  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1              Pharmacy                Coffee Shop                   Pub   
2                  Bank                Pizza Place    Chinese Restaurant   
3                  Bank         Chinese Restaurant   Japanese Restaurant   
4        Sandwich Place         Italian Restaurant              Pharmacy   
5              Beer Bar             Farmers Market    Seafood Restaurant   
6                  Café      General Entertainment          Skating Rink   
7             Nightclub             Breakfast Spot           Coffee Shop   
8         Auto Workshop                 Skate Park                  Park   
9       Airport Service                      Plane       Harbor / Marina   

         5th Most Common Venue 6th Most Common Venue  \
0               Clothing Store          Dessert Shop   
1               Sandwich Place                   Gym   
2                  Bridal Shop         Shopping Mall   
3                  Yoga Studio     Electronics Store   
4              Thai Restaurant           Pizza Place   
5                 Cocktail Bar     French Restaurant   
6  Eastern European Restaurant   Dumpling Restaurant   
7                          Gym               Stadium   
8           Light Rail Station         Burrito Place   
9                  Coffee Shop   Rental Car Location   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
1                      Gay Bar            Donut Shop              Dive Bar   
2               Sandwich Place                 Diner         Deli / Bodega   
3  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
4                          Pub            Restaurant                  Café   
5           Basketball Stadium             Jazz Club           Fish Market   
6                    Drugstore            Donut Shop               Dog Run   
7                          Bar    Italian Restaurant                Bakery   
8               Farmers Market  Fast Food Restaurant               Butcher   
9             Sculpture Garden         Boat or Ferry                   Bar   

  10th Most Common Venue  
0             Donut Shop  
1    Distribution Center  
2             Restaurant  
3             Donut Shop  
4                Butcher  
5                 Bakery  
6               Dive Bar  
7           Intersection  
8             Restaurant  
9          Historic Site

### Cluster Neighbourhoods
Use k-means to cluster the neighbourhood into 5 clusters.

In [68]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)
kmeans
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:150]


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0,
       0, 0, 3, 0, 0, 4, 3, 0, 0, 0, 3, 0, 0, 0, 2, 0, 0, 3, 0, 0, 0, 3,
       0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       3, 0, 0, 0, 0, 0, 3, 1], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighbourhood.

In [72]:

#neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighbourhoods_venues_sorted.head()


Cluster Labels                                    Neighbourhood  \
0               0                                        Agincourt   
1               0                           Alderwood, Long Branch   
2               0  Bathurst Manor, Wilson Heights, Downsview North   
3               0                                  Bayview Village   
4               0                Bedford Park, Lawrence Manor East   

  1st Most Common Venue 2nd Most Common Venue      3rd Most Common Venue  \
0                Lounge        Breakfast Spot  Latin American Restaurant   
1           Pizza Place              Pharmacy                Coffee Shop   
2           Coffee Shop                  Bank                Pizza Place   
3                  Café                  Bank         Chinese Restaurant   
4           Coffee Shop        Sandwich Place         Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0          Skating Rink        Clothing Store          Dessert Shop   
1                   Pub        Sandwich Place                   Gym   
2    Chinese Restaurant           Bridal Shop         Shopping Mall   
3   Japanese Restaurant           Yoga Studio     Electronics Store   
4              Pharmacy       Thai Restaurant           Pizza Place   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
1                      Gay Bar            Donut Shop              Dive Bar   
2               Sandwich Place                 Diner         Deli / Bodega   
3  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
4                          Pub            Restaurant                  Café   

  10th Most Common Venue  
0             Donut Shop  
1    Distribution Center  
2             Restaurant  
3             Donut Shop  
4                Butcher

In [74]:
toronto_merged = df_Toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighbourhood

toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
toronto_merged.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Malvern, Rouge  43.806686   
1      M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.194353             4.0        Fast Food Restaurant   
1 -79.160497             2.0  Construction & Landscaping   
2 -79.188711             0.0           Electronics Store   
3 -79.216917             0.0                 Coffee Shop   
4 -79.239476             0.0        Caribbean Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Yoga Studio  Ethiopian Restaurant     Electronics Store   
1                    Bar           Yoga Studio    Dim Sum Restaurant   
2         Breakfast Spot            Restaurant   Rental Car Location   
3  Korean BBQ Restaurant    Mexican Restaurant           Yoga Studio   
4                   Bank                Bakery    Athletics & Sports   

         5th Most Common Venue 6th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant   
1                  Escape Room     Electronics Store   
2           Mexican Restaurant        Medical Center   
3                 Dessert Shop     Electronics Store   
4              Thai Restaurant           Gas Station   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                    Drugstore            Donut Shop               Dog Run   
1  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
2                 Intersection                  Bank        Discount Store   
3  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
4             Hakka Restaurant   Fried Chicken Joint             Drugstore   

  10th Most Common Venue  
0               Dive Bar  
1             Donut Shop  
2                  Diner  
3             Donut Shop  
4             Donut Shop

Get read of  the NaN values to prevent data skew

In [75]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting Clusters on the map

In [78]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

### Examine Clusters
Let's see each cluster and determine venue categories that distinguish each cluster.

In [80]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, 
                         toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Scarborough             0.0     Electronics Store   
3    Scarborough             0.0           Coffee Shop   
4    Scarborough             0.0  Caribbean Restaurant   
5    Scarborough             0.0            Playground   
6    Scarborough             0.0            Hobby Shop   
..           ...             ...                   ...   
96    North York             0.0           Pizza Place   
99     Etobicoke             0.0           Pizza Place   
100    Etobicoke             0.0           Pizza Place   
101    Etobicoke             0.0         Grocery Store   
102    Etobicoke             0.0             Drugstore   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2            Breakfast Spot            Restaurant   Rental Car Location   
3     Korean BBQ Restaurant    Mexican Restaurant           Yoga Studio   
4                      Bank                Bakery    Athletics & Sports   
5                Smoke Shop         Jewelry Store      Department Store   
6               Coffee Shop      Department Store         Train Station   
..                      ...                   ...                   ...   
96   Furniture / Home Store           Yoga Studio         Deli / Bodega   
99       Chinese Restaurant        Sandwich Place           Coffee Shop   
100                    Park              Bus Line        Sandwich Place   
101             Pizza Place   Fried Chicken Joint        Sandwich Place   
102                     Bar         Garden Center   Rental Car Location   

           5th Most Common Venue        6th Most Common Venue  \
2             Mexican Restaurant               Medical Center   
3                   Dessert Shop            Electronics Store   
4                Thai Restaurant                  Gas Station   
5    Eastern European Restaurant          Dumpling Restaurant   
6                    Yoga Studio           Dim Sum Restaurant   
..                           ...                          ...   
96   Eastern European Restaurant          Dumpling Restaurant   
99                Discount Store                 Intersection   
100             Department Store  Eastern European Restaurant   
101                   Beer Store         Fast Food Restaurant   
102                  Yoga Studio             Department Store   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
2                   Intersection                  Bank        Discount Store   
3    Eastern European Restaurant   Dumpling Restaurant             Drugstore   
4               Hakka Restaurant   Fried Chicken Joint             Drugstore   
5                      Drugstore            Donut Shop               Dog Run   
6                          Diner        Discount Store   Distribution Center   
..                           ...                   ...                   ...   
96                     Drugstore            Donut Shop               Dog Run   
99                  Dessert Shop    Dim Sum Restaurant                 Diner   
100          Dumpling Restaurant             Drugstore            Donut Shop   
101                 Liquor Store              Pharmacy          Dance Studio   
102  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

    10th Most Common Venue  
2                    Diner  
3               Donut Shop  
4               Donut Shop  
5                 Dive Bar  
6                 Dive Bar  
..                     ...  
96                Dive Bar  
99                Dive Bar  
100                Dog Run  
101             Donut Shop  
102                Dog Run  

[86 rows x 12 columns]

In [96]:
a = toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].shape

In [86]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
20  North York             1.0   Martial Arts School           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
20         Grocery Store     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
20   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
20               Dog Run               Dive Bar

In [95]:
b = toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].shape

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1   Scarborough             2.0  Construction & Landscaping   
91    Etobicoke             2.0  Construction & Landscaping   
97   North York             2.0              Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1                    Bar           Yoga Studio    Dim Sum Restaurant   
91        Baseball Field           Yoga Studio    Dim Sum Restaurant   
97           Yoga Studio           Event Space           Escape Room   

   5th Most Common Venue        6th Most Common Venue  \
1            Escape Room            Electronics Store   
91           Escape Room            Electronics Store   
97     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1   Eastern European Restaurant   Dumpling Restaurant             Drugstore   
91  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
97          Dumpling Restaurant             Drugstore            Donut Shop   

   10th Most Common Venue  
1              Donut Shop  
91             Donut Shop  
97                Dog Run

In [94]:
c = toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].shape

In [83]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             3.0            Playground   
23        North York             3.0                  Park   
25        North York             3.0                  Park   
40         East York             3.0                  Park   
44   Central Toronto             3.0                  Park   
50  Downtown Toronto             3.0                  Park   
74              York             3.0                  Park   
79        North York             3.0                  Park   
98              York             3.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                        Park                Bakery          Intersection   
23           Convenience Store           Yoga Studio      Department Store   
25           Food & Drink Shop           Yoga Studio      Department Store   
40           Convenience Store          Intersection           Yoga Studio   
44                 Swim School              Bus Line           Yoga Studio   
50                Tennis Court                 Trail            Playground   
74               Women's Store                  Pool         Deli / Bodega   
79  Construction & Landscaping                Bakery           Yoga Studio   
98                 Yoga Studio      Department Store     Electronics Store   

          5th Most Common Venue        6th Most Common Venue  \
14             Department Store  Eastern European Restaurant   
23            Electronics Store  Eastern European Restaurant   
25  Eastern European Restaurant          Dumpling Restaurant   
40                 Dessert Shop            Electronics Store   
44                 Dessert Shop            Electronics Store   
50                 Dessert Shop           Dim Sum Restaurant   
74  Eastern European Restaurant          Dumpling Restaurant   
79                 Dessert Shop            Electronics Store   
98  Eastern European Restaurant          Dumpling Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14          Dumpling Restaurant             Drugstore            Donut Shop   
23          Dumpling Restaurant             Drugstore            Donut Shop   
25                    Drugstore            Donut Shop               Dog Run   
40  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
44  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
50                        Diner        Discount Store   Distribution Center   
74                    Drugstore            Donut Shop               Dog Run   
79  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
98                    Drugstore            Donut Shop               Dog Run   

   10th Most Common Venue  
14                Dog Run  
23                Dog Run  
25               Dive Bar  
40             Donut Shop  
44             Donut Shop  
50                Dog Run  
74               Dive Bar  
79             Donut Shop  
98               Dive Bar

In [93]:
d = toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].shape


In [84]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             4.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Dumpling Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run               Dive Bar

In [92]:
e = toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].shape

In [116]:
print ('Cluster 1 has',a[0], 'venues')
print ('Cluster 2 has',b[0], 'venues') 
print ('Cluster 3 has',c[0], 'venues') 
print ('Cluster 4 has',d[0], 'venues') 
print ('Cluster 5 has',e[0], 'venues') 

Cluster 1 has 86 venues
Cluster 2 has 1 venues
Cluster 3 has 3 venues
Cluster 4 has 9 venues
Cluster 5 has 9 venues


We clustered Toronto neighbourhood based on venue categories.